src/components/Microsoft.DotNet.UpgradeAssistant/TargetFramework/TargetFrameworkSelector.cs

```diff
@@ -0,0 +1,100 @@
+
+        private readonly string _currentTFMBase;
+        private readonly string _ltsTFMBase;
+
+        private readonly UpgradeTarget _upgradeTarget;
+
+        public TargetFrameworkSelector(UpgradeOptions options, ITargetFrameworkMonikerComparer comparer, IOptions<TFMSelectorOptions> selectorOptions, IEnumerable<ITargetFrameworkSelectorFilter> selectors)
+        {
+            _comparer = comparer;
+            _selectors = selectors;
+
+            _currentTFMBase = selectorOptions?.Value.CurrentTFMBase ?? DefaultCurrentTFMBase;
+            _ltsTFMBase = selectorOptions?.Value.LTSTFMBase ?? DefaultLTSTFMBase;
+            _upgradeTarget = options?.UpgradeTarget ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        public async ValueTask<TargetFrameworkMoniker> SelectTargetFrameworkAsync(IProject project, CancellationToken token)
+        {
+            if (project is null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            var appBase = _upgradeTarget == UpgradeTarget.Current ? _currentTFMBase : _ltsTFMBase;
+            var current = GetDefaultTargetFrameworkMoniker(project);
+            var appBaseTfm = new TargetFrameworkMoniker(appBase);
+
+            var updater = new FilterState(_comparer, project, current, appBaseTfm)
+            {
+                Components = await project.GetComponentsAsync(token).ConfigureAwait(false),
+            };
+
+            foreach (var selector in _selectors)
+            {
+                selector.Process(updater);
+            }
+
+            return updater.Current;
+        }
+
+        private static TargetFrameworkMoniker GetDefaultTargetFrameworkMoniker(IProject project)
+        {
+            foreach (var currentTfm in project.TargetFrameworks)
+            {
+                if (currentTfm.IsNetStandard)
+                {
+                    return currentTfm;
+                }
+            }
+
+            return TargetFrameworkMoniker.NetStandard20;
+        }
+
+        private class FilterState : ITargetFrameworkSelectorFilterState
+        {
+            private readonly ITargetFrameworkMonikerComparer _comparer;
+
+            public FilterState(ITargetFrameworkMonikerComparer comparer, IProject project, TargetFrameworkMoniker current, TargetFrameworkMoniker appbase)
+            {
+                _comparer = comparer;
+                Project = project;
+                Current = current;
+                AppBase = appbase;
+            }
+
+            public TargetFrameworkMoniker Current { get; private set; }
+
+            public TargetFrameworkMoniker AppBase { get; }
+
+            public IProject Project { get; }
+
+            public ProjectComponents Components { get; init; }
+
+            public bool TryUpdate(TargetFrameworkMoniker tfm)
+            {
+                Current = _comparer.Merge(Current, tfm);
+
+                return Current != tfm;
```

Comment:
Should we check Current != tfm before merging? If Current starts as net5.0 and tfm is net5.0-windows, then after merging they'll be the same, but an update did happen.

Reply:
Good catch!